src/ccmain/thresholder.cpp

```diff
@@ -184,6 +185,45 @@ void ImageThresholder::SetImage(const Image pix) {
   Init();
 }
 
+std::tuple<bool, Image, Image, Image> ImageThresholder::Threshold(
+                                                         ThreshMethod method) {
+  Image pix_grey = nullptr;
+  Image pix_binary = nullptr;
+  Image pix_thresholds = nullptr;
+
+  if (image_width_ > INT16_MAX || image_height_ > INT16_MAX) {
+    tprintf("Image too large: (%d, %d)\n", image_width_, image_height_);
+    return std::make_tuple(false, nullptr, nullptr, nullptr);
+  }
+  
+  if (pix_channels_ == 0) {
+    // We have a binary image, but it still has to be copied, as this API
+    // allows the caller to modify the output.
+    Image original = GetPixRect();
+    pix_binary = original.copy();
+    original.destroy();
+    return std::make_tuple(false, nullptr, pix_binary, nullptr);
+  }
+
+  pix_grey = GetPixRectGrey();
+
+  if (method == ThreshMethod::Otsu || method >= ThreshMethod::Count) {
+    method = ThreshMethod::AdaptiveOtsu;
+  }
+
+  int r;
+  if (method == ThreshMethod::AdaptiveOtsu) {
+    r = pixOtsuAdaptiveThreshold(pix_grey, 300, 300, 0, 0, 0.1, 
+                                 pix_thresholds.a(), pix_binary.a());
+  } else if (method == ThreshMethod::TiledSauvola) {
+    r = pixSauvolaBinarizeTiled(pix_grey, 25, 0.40, 300, 300, pix_thresholds.a(), 
+                                pix_binary.a());
```

Comment:
@amitdo I am perplexed by this. According to Leptonica's API documentation, the 4th and 5th parameters are not tile **size** but tile **number** – so you don't get 300 but `image_width_ / 300` pixels here. 

Also, the _window size_ parameter of 25 seems **too low** for typical cases (like 300 DPI input) – remember that in the Sauvola formula, this needs to encompass enough statistics about each pixels environment to include both enough background and foreground. (I have found a good rule of thumb to be "window size equals DPI", because one inch of print usually contains enough whitespace around even large letters. I did experiment a lot with this myself, but cannot say much about the literature on the subject.)

Finally, I think the `factor=0.4` (which is the `k` parameter in the Sauvola formula, and usually is described as 0.34) should really be exposed as a Tesseract parameter.

Reply:
HI @bertsky,  

>According to Leptonica's API documentation, the 4th and 5th parameters are not tile **size** but tile **number** – so you don't get 300 but `image_width_ / 300` pixels here.

Yes, you're right (but Leptonica is smart enough to correct my mistake in most cases). 

What should be the value here?

In another function, `pixSauvolaOnContrastNorm()` Leptonica has this code:

```
     /* Use tiles of size approximately 250 x 250 */
    pixGetDimensions(pix1, &w, &h, NULL);
    nx = L_MAX(1, (w + 125) / 250);
    ny = L_MAX(1, (h + 125) / 250);
    pixSauvolaBinarizeTiled(pix1, 25, 0.40, nx, ny, ppixth, &pixd);
```

Reply:
> Yes, you're right (but Leptonica is smart enough to correct my mistake in most cases).

Indeed, Leptonica does some range validation:
```C
    xrat = w / nx;
    yrat = h / ny;
    if (xrat < whsize + 2) {
        nx = w / (whsize + 2);
        L_WARNING("tile width too small; nx reduced to %d\n", procName, nx);
    }
    if (yrat < whsize + 2) {
        ny = h / (whsize + 2);
        L_WARNING("tile height too small; ny reduced to %d\n", procName, ny);
    }
```
So tiles will always be at least 1px larger than windows. (For us, of course, window size is equally problematic, as sketched above.)

> What should be the value here?
> 
> In another function, `pixSauvolaOnContrastNorm()` Leptonica has this code:

Hm, so they want to 250px tiles. That's not going to work if DPI is very large (say 1200) and therefore windows also need to be larger.

Since tile size is only about performance (memory / runtime trade-off), and window size only about quality (fg / bg statistics), I propose:

1. Use `estimated_res_` for DPI. (If input DPI is implausible, this will have been overriden by DPI estimation. In any case, the user can override DPI here as well.)
2. Set window size by multiplying DPI with a variable factor, let's say `thresholding_win_scale=0.5`. (The user can override that if she has a-priori knowledge about the material, e.g. 0.2 for small font size pages or 0.9 for title pages.)
3. Set tile size by dividing image size with a variable divisor, let's say `thresholding_tiles=8`, but ensure it is between window size and 1.

Reply:
Of course, we should first experiment on all these parameters with a suitable GT. Assuming we can definitely rule out any effects of binarization on neural recognition, all we need is layout evaluation. We could use PRImA tools for this on structural GT, or simple pixelwise mAP scoring...

Reply:
#3570 fixed the wrong tiles parameters in Sauvola.